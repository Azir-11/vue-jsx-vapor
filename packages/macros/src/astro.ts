import unplugin from '.'
import type { Options } from './options'

export default (options: Options) => ({
  name: 'unplugin-vue-jsx-vapor',
  hooks: {
    'astro:config:setup': (astro: any) => {
      astro.config.vite.plugins ||= []
      astro.config.vite.plugins.push(unplugin.vite(options))
    },
  },
})
