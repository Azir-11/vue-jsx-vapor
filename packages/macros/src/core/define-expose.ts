import { importHelperFn, type MagicStringAST } from '@vue-macros/common'
import type { CallExpression } from '@babel/types'

export function transformDefineExpose(
  node: CallExpression,
  s: MagicStringAST,
  lib: string,
): void {
  s.overwriteNode(node.callee, '')
  s.appendRight(
    node.arguments[0]?.start || node.end! - 1,
    lib.includes('vapor')
      ? `${importHelperFn(s, 0, 'currentInstance', 'vue')}.exposed = `
      : `${importHelperFn(s, 0, 'getCurrentInstance', 'vue')}().exposed = `,
  )
}
