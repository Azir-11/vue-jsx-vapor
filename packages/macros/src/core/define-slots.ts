import { importHelperFn, type MagicStringAST } from '@vue-macros/common'
import type { CallExpression } from '@babel/types'

export function transformDefineSlots(
  node: CallExpression,
  s: MagicStringAST,
): void {
  s.overwrite(
    node.start!,
    (node.arguments[0]?.start && node.arguments[0].start - 1) ||
      node.typeArguments?.end ||
      node.callee.end!,
    `Object.assign`,
  )
  const slots = `${importHelperFn(s, 0, 'useSlots')}()`
  s.appendLeft(node.end! - 1, `${node.arguments[0] ? ',' : '{}, '}${slots}`)
}
