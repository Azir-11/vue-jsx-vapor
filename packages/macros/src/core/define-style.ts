import { walkAST, type MagicStringAST } from '@vue-macros/common'
import hash from 'hash-sum'
import { helperPrefix } from './helper'
import type { FunctionalNode } from './babel-utils'
import { isFunctionalNode, type DefineStyle, type Macros } from '.'
import type { Node } from '@babel/types'

export function transformDefineStyle(
  defineStyle: DefineStyle,
  index: number,
  root: FunctionalNode | undefined,
  s: MagicStringAST,
  importMap: Map<string, string>,
  { defineSlots }: Macros,
): void {
  const { expression, lang, isCssModules } = defineStyle
  if (expression.arguments[0]?.type !== 'TemplateLiteral') return

  let css = s.sliceNode(expression.arguments[0]).slice(1, -1)
  const scopeId = hash(css)
  const vars = new Map<string, string>()
  expression.arguments[0].expressions.forEach((exp) => {
    const cssVar = s.sliceNode(exp)
    const cssVarId = toCssVarId(cssVar, `--${scopeId}-`)
    s.overwrite(exp.start! - 2, exp.end! + 1, `var(${cssVarId})`)
    vars.set(cssVarId, cssVar)
  })

  let returnExpression = root && getReturnStatement(root)
  if (isFunctionalNode(returnExpression)) {
    returnExpression = getReturnStatement(returnExpression)
  }
  if (vars.size && returnExpression) {
    const children =
      returnExpression.type === 'JSXElement'
        ? [returnExpression]
        : returnExpression.type === 'JSXFragment'
          ? returnExpression.children
          : []
    const varString = Array.from(vars.entries())
      .map(([key, value]) => `'${key}': ${value}`)
      .join(', ')
    for (const child of children) {
      if (child.type === 'JSXElement') {
        s.appendRight(
          child.openingElement.name.end!,
          ` {...{style:{${varString}}}}`,
        )
      }
    }
  }

  let scoped = !!root
  if (expression.arguments[1]?.type === 'ObjectExpression') {
    for (const prop of expression.arguments[1].properties) {
      if (
        prop.type === 'ObjectProperty' &&
        prop.key.type === 'Identifier' &&
        prop.key.name === 'scoped' &&
        prop.value.type === 'BooleanLiteral'
      ) {
        scoped = prop.value.value
      }
    }
  }

  if (scoped && root) {
    const slotNames = defineSlots?.id
      ? defineSlots.id.type === 'Identifier'
        ? defineSlots.id.name
        : defineSlots.id.type === 'ObjectPattern'
          ? defineSlots.id.properties.map((prop) =>
              s.sliceNode(
                prop.type === 'RestElement' ? prop.argument : prop.value,
              ),
            )
          : []
      : []
    walkAST<Node>(root, {
      enter(node) {
        if (
          node.type === 'JSXElement' &&
          s.sliceNode(node.openingElement.name) !== 'template'
        ) {
          let subfix = ''
          if (slotNames.length) {
            const name = s.sliceNode(
              node.openingElement.name.type === 'JSXMemberExpression'
                ? node.openingElement.name.object
                : node.openingElement.name,
            )
            subfix = slotNames.includes(name) ? '-s' : ''
          }
          s.appendRight(
            node.openingElement.name.end!,
            ` data-v-${scopeId}${subfix}=""`,
          )
        }
      },
    })
  }

  css = s
    .sliceNode(expression.arguments[0])
    .slice(1, -1)
    .replaceAll(/\/\/(.*)(?=\n)/g, '/*$1*/')
  const module = isCssModules ? 'module.' : ''
  const importId = `${helperPrefix}/define-style/${index}?scopeId=${scopeId}&scoped=${scoped}&lang.${module}${lang}`
  importMap.set(importId, css)
  s.appendLeft(
    0,
    isCssModules
      ? `import style${index} from "${importId}";`
      : `import "${importId}";`,
  )
  s.overwriteNode(expression, isCssModules ? `style${index}` : '')
}

function getReturnStatement(root: FunctionalNode) {
  if (root.body.type === 'BlockStatement') {
    const returnStatement = root.body.body.find(
      (node) => node.type === 'ReturnStatement',
    )
    if (returnStatement) {
      return returnStatement.argument
    }
  } else {
    return root.body
  }
}

function toCssVarId(name: string, prefix = '') {
  return (
    prefix +
    name.replaceAll(/\W/g, (searchValue, replaceValue) => {
      return searchValue === '.'
        ? '-'
        : name.charCodeAt(replaceValue).toString()
    })
  )
}
