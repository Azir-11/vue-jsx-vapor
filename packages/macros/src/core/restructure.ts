import {
  HELPER_PREFIX,
  importHelperFn,
  walkIdentifiers,
  type MagicStringAST,
} from '@vue-macros/common'
import { withDefaultsHelperId } from './helper'
import {
  getDefaultValue,
  prependFunctionalNode,
  type FunctionalNode,
} from './utils'
import type { Node } from '@babel/types'

type Options = {
  withDefaultsFrom?: string
  skipDefaultProps?: boolean
  generateRestProps?: (
    restPropsName: string,
    index: number,
    list: Prop[],
  ) => string | undefined
}

type Prop = {
  path: string
  name: string
  value: string
  defaultValue?: string
  isRest?: boolean
}

export function restructure(
  s: MagicStringAST,
  node: FunctionalNode,
  options: Options = {},
): Prop[] {
  let index = 0
  const propList: Prop[] = []
  for (const param of node.params) {
    const path = `${HELPER_PREFIX}props${index++ || ''}`
    const props = getProps(s, options, param, path)
    if (props) {
      s.overwrite(param.start!, param.end!, path)
      propList.push(...props)
    }
  }

  if (propList.length) {
    const defaultValues: Record<string, Prop[]> = {}
    const rests = []
    for (const prop of propList) {
      if (prop.isRest) {
        rests.push(prop)
      }
      if (prop.defaultValue) {
        const paths = prop.path.split(/\.|\[/)
        if (!options.skipDefaultProps || paths.length !== 1) {
          ;(defaultValues[paths[0]] ??= []).push(prop)
        }
      }
    }

    for (const [index, rest] of rests.entries()) {
      prependFunctionalNode(
        node,
        s,
        options.generateRestProps?.(rest.name, index, rests) ??
          `\nconst ${rest.name} = ${importHelperFn(
            s,
            0,
            'createPropsRestProxy',
          )}(${rest.path}, [${rest.value}])`,
      )
    }

    for (const [path, values] of Object.entries(defaultValues)) {
      const createPropsDefaultProxy = importHelperFn(
        s,
        0,
        'createPropsDefaultProxy',
        undefined,
        options.withDefaultsFrom ?? withDefaultsHelperId,
      )
      const resolvedValues = values
        .map(
          (i) => `'${i.path.replace(path, '')}${i.value}': ${i.defaultValue}`,
        )
        .join(', ')
      prependFunctionalNode(
        node,
        s,
        `\n${path} = ${createPropsDefaultProxy}(${path}, {${resolvedValues}})`,
      )
    }

    walkIdentifiers(
      node.body,
      (id, parent) => {
        const prop = propList.find((i) => i.name === id.name)
        if (prop && !prop.isRest) {
          s.overwrite(
            id.start!,
            id.end!,
            `${
              parent?.type === 'ObjectProperty' && parent.shorthand
                ? `${id.name}: `
                : ''
            }${prop.path}${prop.value}`,
          )
        }
      },
      false,
    )
  }

  return propList
}

function getProps(
  s: MagicStringAST,
  options: Options,
  node: Node,
  path = '',
  props: Prop[] = [],
) {
  const properties =
    node.type === 'ObjectPattern'
      ? node.properties
      : node.type === 'ArrayPattern'
        ? node.elements
        : []
  if (!properties.length) return

  const propNames: string[] = []
  properties.forEach((prop, index) => {
    if (prop?.type === 'Identifier') {
      // { foo }
      props.push({
        name: prop.name,
        path,
        value: `[${index}]`,
      })
      propNames.push(`'${prop.name}'`)
    } else if (
      prop?.type === 'AssignmentPattern' &&
      prop.left.type === 'Identifier'
    ) {
      // [foo = 'foo']
      props.push({
        path,
        name: prop.left.name,
        value: `[${index}]`,
        defaultValue: s.sliceNode(getDefaultValue(prop.right)),
      })
      propNames.push(`'${prop.left.name}'`)
    } else if (
      prop?.type === 'ObjectProperty' &&
      prop.key.type === 'Identifier'
    ) {
      if (prop.value.type === 'AssignmentPattern') {
        if (prop.value.left.type === 'Identifier') {
          // { foo: bar = 'foo' }
          props.push({
            path,
            name: prop.value.left.name,
            value: `.${prop.key.name}`,
            defaultValue: s.sliceNode(getDefaultValue(prop.value.right)),
          })
        } else {
          // { foo: { bar } = {} }
          getProps(
            s,
            options,
            prop.value.left,
            `${path}.${prop.key.name}`,
            props,
          )
        }
      } else if (
        !getProps(s, options, prop.value, `${path}.${prop.key.name}`, props)
      ) {
        // { foo: bar }
        const name =
          prop.value.type === 'Identifier' ? prop.value.name : prop.key.name
        props.push({
          path,
          name,
          value: `.${prop.key.name}`,
        })
      }
      propNames.push(`'${prop.key.name}'`)
    } else if (
      prop?.type === 'RestElement' &&
      prop.argument.type === 'Identifier' &&
      !prop.argument.name.startsWith(`${HELPER_PREFIX}props`)
    ) {
      // { ...rest }
      props.push({
        path,
        name: prop.argument.name,
        value: propNames.join(', '),
        isRest: true,
      })
    } else if (prop) {
      getProps(s, options, prop, `${path}[${index}]`, props)
    }
  })
  return props.length ? props : undefined
}
