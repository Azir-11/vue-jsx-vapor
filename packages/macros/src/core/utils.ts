import type {
  ArrowFunctionExpression,
  FunctionDeclaration,
  FunctionExpression,
  Node,
} from '@babel/types'
import type { MagicStringAST } from '@vue-macros/common'

export type FunctionalNode =
  | FunctionDeclaration
  | FunctionExpression
  | ArrowFunctionExpression

export function prependFunctionalNode(
  node: FunctionalNode,
  s: MagicStringAST,
  result: string,
): void {
  const isBlockStatement = node.body.type === 'BlockStatement'
  const start = node.body.extra?.parenthesized
    ? (node.body.extra.parenStart as number)
    : node.body.start!
  s.appendRight(
    start + (isBlockStatement ? 1 : 0),
    `${result};${!isBlockStatement ? 'return ' : ''}`,
  )
  if (!isBlockStatement) {
    s.appendLeft(start, '{')
    s.appendRight(node.end!, '}')
  }
}

export function isFunctionalNode(node?: Node | null): node is FunctionalNode {
  return !!(
    node &&
    (node.type === 'ArrowFunctionExpression' ||
      node.type === 'FunctionDeclaration' ||
      node.type === 'FunctionExpression')
  )
}

export function getParamsStart(node: FunctionalNode, code: string): number {
  return node.params[0]
    ? node.params[0].start!
    : node.start! +
        (code.slice(node.start!, node.body.start!).match(/\(\s*\)/)?.index ||
          0) +
        1
}

export function getDefaultValue(node: Node): Node {
  if (node.type === 'TSNonNullExpression') {
    return getDefaultValue(node.expression)
  }
  if (node.type === 'TSAsExpression') {
    return getDefaultValue(node.expression)
  }
  return node
}
