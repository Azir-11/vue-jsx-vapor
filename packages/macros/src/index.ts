import { createUnplugin, type UnpluginInstance } from 'unplugin'
import plugin from './raw'
import type { Options } from './options'

export * from './options'

const unplugin: UnpluginInstance<Options | undefined> = createUnplugin(plugin)
export default unplugin
