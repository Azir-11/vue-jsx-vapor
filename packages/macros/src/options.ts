import {
  detectVueVersion,
  REGEX_NODE_MODULES,
  REGEX_SETUP_SFC,
  REGEX_SRC_FILE,
  type BaseOptions,
  type MarkRequired,
} from '@vue-macros/common'

export type Options = BaseOptions & {
  lib?: 'vue' | 'vue/vapor' | (string & {})
  defineComponent?: { alias: string[] }
  defineModel?: { alias: string[] }
  defineExpose?: { alias: string[] }
  defineSlots?: { alias: string[] }
  defineStyle?: { alias: string[] }
}
export type OptionsResolved = MarkRequired<
  Options,
  | 'include'
  | 'version'
  | 'lib'
  | 'defineComponent'
  | 'defineModel'
  | 'defineExpose'
  | 'defineSlots'
  | 'defineStyle'
>

export function resolveOptions(options: Options): OptionsResolved {
  const version = options.version || detectVueVersion()
  const lib = options.lib || 'vue/vapor'
  return {
    include: [REGEX_SRC_FILE],
    exclude: [REGEX_SETUP_SFC, REGEX_NODE_MODULES],
    ...options,
    version,
    lib,
    defineComponent: {
      alias: options?.defineComponent?.alias ?? [
        'defineComponent',
        'defineVaporComponent',
      ],
    },
    defineModel: { alias: options?.defineModel?.alias ?? ['defineModel'] },
    defineSlots: { alias: options?.defineSlots?.alias ?? ['defineSlots'] },
    defineExpose: { alias: options?.defineExpose?.alias ?? ['defineExpose'] },
    defineStyle: { alias: options?.defineStyle?.alias ?? ['defineStyle'] },
  }
}
