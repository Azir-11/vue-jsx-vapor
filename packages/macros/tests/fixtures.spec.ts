import process from 'node:process'
import { testFixtures } from '@vue-macros/test-utils'
import { describe } from 'vitest'
import { transformJsxMacros } from '../src/core'

const options = {
  defineModel: { alias: ['defineModel'] },
  defineSlots: { alias: ['defineSlots'] },
  defineStyle: { alias: ['defineStyle'] },
  defineExpose: { alias: ['defineExpose'] },
  defineComponent: { alias: ['defineComponent', 'defineVaporComponent'] },
}

// TODO: hash-sum's result is different on Windows and Linux
const globs =
  process.platform === 'win32'
    ? import.meta.glob(
        ['./fixtures/**/*.tsx', '!./fixtures/**/define-style.tsx'],
        {
          eager: true,
          as: 'raw',
        },
      )
    : import.meta.glob('./fixtures/**/*.tsx', {
        eager: true,
        as: 'raw',
      })

describe('fixtures', async () => {
  await testFixtures(
    globs,
    (args, id, code) =>
      transformJsxMacros(code, id, new Map(), {
        lib: 'vue',
        include: ['*.tsx'],
        version: 3.6,
        ...options,
      })?.code,
  )
})

describe('vue/vapor fixtures', async () => {
  await testFixtures(
    globs,
    (args, id, code) =>
      transformJsxMacros(code, id, new Map(), {
        lib: 'vue/vapor',
        include: ['*.tsx'],
        version: 3.6,
        ...options,
      })?.code,
  )
})
