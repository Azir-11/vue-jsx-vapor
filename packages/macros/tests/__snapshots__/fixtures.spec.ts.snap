// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`defineComponent autoReturnFunction fixtures > ./fixtures/define-component.tsx 1`] = `
"
import { createPropsDefaultProxy as __MACROS_createPropsDefaultProxy } from "vue-jsx-vapor/macros/with-defaults";
import { useAttrs as __MACROS_useAttrs } from "vue";
import { useModel as __MACROS_useModel } from "vue-jsx-vapor/macros/use-model";
import { withAsyncContext as __MACROS_withAsyncContext } from "vue";import { defineComponent, nextTick } from 'vue'

const Comp = defineComponent(
  (__MACROS_props) => {
const __MACROS_default_props = __MACROS_createPropsDefaultProxy(__MACROS_props, {'.bar': 'bar'!});const attrs = __MACROS_useAttrs();
    __MACROS_useModel(__MACROS_props, 'modelValue',)
    const foo = $(
      __MACROS_useModel(__MACROS_props, 'foo', {
        validator: (value) => {
          return value === 'foo'
        },
        type: String,
      })!,
    )
    return () => <div>{[foo, __MACROS_default_props.bar, attrs.baz]}</div>
  },
  {props: { 'bar': { required: true }, 'modelValue': null, 'onUpdate:modelValue': null, 'modelModifiers': null, 'foo': { required: true, validator: (value) => {
          return value === 'foo'
        }, type: String }, 'onUpdate:foo': null, 'fooModifiers': null },inheritAttrs: false, name: 'Comp' },
)

const Comp1 = defineComponent((props: { bar: 'bar'; 'onUpdate:bar': any }) => {
  const foo = __MACROS_useModel(props, 'foo')
  return () => <div>{[foo.value, props['bar'], props['onUpdate:bar']]}</div>
}, { props: { 'bar': null, 'onUpdate:bar': null, 'foo': null, 'onUpdate:foo': null, 'fooModifiers': null } })

const Comp2 = defineComponent(async (__MACROS_props) => {
let __temp, __restore

  ;(
  ([__temp,__restore] = __MACROS_withAsyncContext(() => nextTick())),
  await __temp,
  __restore()
)
  let foo = (
  ([__temp,__restore] = __MACROS_withAsyncContext(() => new Promise((resolve) => {
    setTimeout(() => resolve('foo'), 1000)
  }))),
  __temp = await __temp,
  __restore(),
  __temp
)
  return () => <div>{foo}</div>
})
"
`;

exports[`fixtures > ./fixtures/define-component.tsx 1`] = `
"
import { createPropsDefaultProxy as __MACROS_createPropsDefaultProxy } from "vue-jsx-vapor/macros/with-defaults";
import { useAttrs as __MACROS_useAttrs } from "vue";
import { useModel as __MACROS_useModel } from "vue-jsx-vapor/macros/use-model";
import { withAsyncContext as __MACROS_withAsyncContext } from "vue";import { defineComponent, nextTick } from 'vue'

const Comp = defineComponent(
  (__MACROS_props) => {
const __MACROS_default_props = __MACROS_createPropsDefaultProxy(__MACROS_props, {'.bar': 'bar'!});const attrs = __MACROS_useAttrs();
    __MACROS_useModel(__MACROS_props, 'modelValue',)
    const foo = $(
      __MACROS_useModel(__MACROS_props, 'foo', {
        validator: (value) => {
          return value === 'foo'
        },
        type: String,
      })!,
    )
    return <div>{[foo, __MACROS_default_props.bar, attrs.baz]}</div>
  },
  {props: { 'bar': { required: true }, 'modelValue': null, 'onUpdate:modelValue': null, 'modelModifiers': null, 'foo': { required: true, validator: (value) => {
          return value === 'foo'
        }, type: String }, 'onUpdate:foo': null, 'fooModifiers': null },inheritAttrs: false, name: 'Comp' },
)

const Comp1 = defineComponent((props: { bar: 'bar'; 'onUpdate:bar': any }) => {
  const foo = __MACROS_useModel(props, 'foo')
  return <div>{[foo.value, props['bar'], props['onUpdate:bar']]}</div>
}, { props: { 'bar': null, 'onUpdate:bar': null, 'foo': null, 'onUpdate:foo': null, 'fooModifiers': null } })

const Comp2 = defineComponent(async (__MACROS_props) => {
let __temp, __restore

  ;(
  ([__temp,__restore] = __MACROS_withAsyncContext(() => nextTick())),
  await __temp,
  __restore()
)
  let foo = (
  ([__temp,__restore] = __MACROS_withAsyncContext(() => new Promise((resolve) => {
    setTimeout(() => resolve('foo'), 1000)
  }))),
  __temp = await __temp,
  __restore(),
  __temp
)
  return () => <div>{foo}</div>
})
"
`;

exports[`fixtures > ./fixtures/define-expose.tsx 1`] = `
"
import { currentInstance as __MACROS_currentInstance } from "vue";export function Comp(__MACROS_props) {
  ;(__MACROS_currentInstance.exposed = {
    foo: 1,
  })
  return <div />
}

export const Comp1 = function (props: any) {
  ;(__MACROS_currentInstance.exposed = {
    foo: props.foo,
  })
  return <div />
}

export const Comp2 = ({ foo, ...__MACROS_props }: any) => {
  ;(__MACROS_currentInstance.exposed = {
    foo,
  })
  return <div />
}
"
`;

exports[`fixtures > ./fixtures/define-model.tsx 1`] = `
"
import { useModel as __MACROS_useModel } from "vue-jsx-vapor/macros/use-model";export const Comp = ({ bar, ...__MACROS_props }: { bar: string }) => {
  const foo = __MACROS_useModel<string, 'm1' | 'm2'>(__MACROS_props, 'foo', { default: bar })
  return <div>{foo.value}</div>
}

export default function (__MACROS_props) {
  const modelValue = $(__MACROS_useModel<string>(__MACROS_props, 'modelValue',)!)
  return (
    <Comp v-model:foo_m1={modelValue} bar="bar">
      {modelValue}
    </Comp>
  )
}
"
`;

exports[`fixtures > ./fixtures/define-slots.tsx 1`] = `
"
import { useSlots as __MACROS_useSlots } from "vue";export const Comp = (__MACROS_props) => {
  const slots = Object.assign<{
    default: () => any
  }>({}, __MACROS_useSlots())
  return <div>{slots.default?.()}</div>
}

export default function (__MACROS_props) {
  const slots = Object.assign({
    default: () => <div>default</div>,
  },__MACROS_useSlots())
  return <div>{slots.default?.()}</div>
}
"
`;

exports[`fixtures > ./fixtures/define-style.tsx 1`] = `
"import "vue-jsx-vapor/macros/define-style/0?scopeId=4e9d5cd0&scoped=true&lang.css";import style1 from "vue-jsx-vapor/macros/define-style/1?scopeId=426a859d&scoped=true&lang.module.scss";import "vue-jsx-vapor/macros/define-style/2?scopeId=63a7910c&scoped=false&lang.scss";import { defineComponent, ref } from 'vue'

export const Comp = (__MACROS_props) => {
  const color = ref('red')
  
  return <div {...{style:{'--4e9d5cd0-color-value': color.value}}} data-v-4e9d5cd0="" class="foo">foo</div>
}

export default defineComponent((__MACROS_props) => {
  const color = ref('red')
  const styles = style1
  return () => (
    <>
      <div {...{style:{'--426a859d-color-value': color.value}}} data-v-426a859d="" class={styles.bar}>foo</div>
      <div {...{style:{'--426a859d-color-value': color.value}}} data-v-426a859d="" class="bar">
        <span data-v-426a859d="">bar</span>
      </div>
    </>
  )
})


"
`;
